game.py

```diff
@@ -0,0 +1,908 @@
+            if any_moves:
+                break
+            for row in range(rows):
+                if self._board[row][column_order[column]] == None:
+                    moves.append ((row, column_order[column]))
+                    break
+                
+        return moves
+   
+    def game_end(self):
+
+# This determines whether the game has ended after the last move and whether it is a win, lose
+# or draw (if the board is full)
+
+        if self.get_number_of_moves() == 0:
+            return False, False        
+        player = - self.get_player_next()
+        board = self.get_board()
+        connect_number = self.get_connect_number()
+        last_move = self.get_last_move()
+
+        global rows
+        global columns
+        
+        game_end = False
+        win = False
+        
+        vectors = [(0, 1), (1, 0), (1, 1), (1, -1)]
+       
+        if last_move == None:
+            return game_end, win
+                     
+        for vector in vectors:
+            consec = 1
+            next_position = last_move
+            while not win:
+                next_position = (next_position[0] + vector[0], next_position[1] + vector[1])
+                # print('Next position positive : ', next_position)
+                if next_position[0] not in range(rows) or next_position[1] not in range(columns):
+                    break                                                                                    
+                if board[next_position[0]][next_position[1]] == player:
+                    consec = consec + 1
+                else:
+                    break
+                if consec == connect_number:
+                        win = True
+                # print('consec positive : ', consec, 'Vector : ', vector)
+                                        
+            next_position = last_move
+            while not win:
+                next_position = (next_position[0] - vector[0], next_position[1] - vector[1])
+                # print('Next position negative : ', next_position)
+                if next_position[0] not in range(rows) or next_position[1] not in range(columns):
+                    break                                                                                    
+                if board[next_position[0]][next_position[1]] == player:
+                    consec = consec + 1
+                else:
+                    break
+                if consec == connect_number:
+                    win = True
+                # print('consec negative : ', consec, 'Vector : ', vector)
+            
+        if win:
+            game_end = True
+ 
+        if None not in (item for sublist in self.get_board() for item in sublist):
+            game_end = True
+        
+        return game_end, win
+
+    def valid_move(self, move):
+        if move in self.get_next_moves():
+            return True
+        else:
+            return False
+      
+    def score(self):

```

Comment:
More long methods here that I think you're wondering whether they should be on/off of this class.

Refactoring is not easy, especially being fresh-eyed to the code, but some thoughts that are relevant:
* Think and review state-dependencies of each of these functions i.e. when does it use `self` and then review that across your methods to get a better sense of that topology, and how it might be broken up into groups (maybe hierarchical, maybe not).
* Now is the best time to refactor! It seems you are close to moving off this project, at least temporarily. You've also had a lot of experience getting deep in this code and changing it recently. This both gives you a deep understanding of how all this code is connected beyond how its laid out in functions (which you'll lose quickly) but also you have experimented with making a lot of changes and how a good idea about which parts you'd like to keep easy-to-change.
* Every time you have a gap and a comment in this function, could it be split into a separate function that just has its results passed through to the next section? Doing it at every point might not work, but I'm sure some will. If you attempt to write unit-tests for this code, I think it will become apparent how much easier it is to assert correctness of smaller composable chunks of code.
* Making this code readable is not an easy task, so don't fret too much. Determining contiguous lines in a grid isn't easy

Reply:
I have access to ChatGPT4, so for fun I asked it to code this up for us. My ask below: 

Please write me Python code to determine how many "open threats" I have in a game of Connect-4. Here, I represent the Connect-4 grid as a list-of-lists, and each cell either has a tile, from player 1, from player 2, or is empty. I define an "open threat" as having three tiles in-a-row for a player (the tiles can be horizontal, vertical or diagonal) and with the fourth tile currently empty.

Reply:
You can use the following Python code to determine the number of open threats in a Connect-4 game:

```
def count_open_threats(grid, player):
    rows, cols = len(grid), len(grid[0])
    directions = [(1, 0), (0, 1), (1, 1), (1, -1)]
    open_threats = 0

    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == player:
                for dr, dc in directions:
                    count = 1
                    for i in range(1, 4):
                        nr, nc = r + dr * i, c + dc * i
                        if 0 <= nr < rows and 0 <= nc < cols:
                            if grid[nr][nc] == player:
                                count += 1
                            elif grid[nr][nc] == 0:
                                if count == 3:
                                    open_threats += 1
                                break
                            else:
                                break
                        else:
                            break

    return open_threats

# Example grid
grid = [
    [1, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0],
    [1, 2, 2, 2, 0, 0, 0],
    [2, 1, 1, 1, 0, 0, 0],
    [2, 2, 2, 1, 0, 0, 0],
    [1, 1, 2, 2, 0, 0, 0]
]

player1_open_threats = count_open_threats(grid, 1)
player2_open_threats = count_open_threats(grid, 2)

print("Player 1 open threats:", player1_open_threats)
print("Player 2 open threats:", player2_open_threats)
```

This code defines a function count_open_threats that takes a grid and a player number (1 or 2) as input and returns the number of open threats for that player. The grid is represented as a list of lists, with 0 for empty cells, 1 for player 1's tiles, and 2 for player 2's tiles.

The code iterates through each cell in the grid, and if the cell contains the given player's tile, it checks for open threats in all four directions (horizontal, vertical, and both diagonals). It keeps track of the number of open threats and returns the final count.

Reply:
^This could be the start of an iterative conversion, for example I followed up with

I don't think this is re-initializing the `count` variable correctly. We are trying to only count an open-threat if a player has 3 tiles in-a-row in a single direction, and that the empty tile is then in that same direction

Reply:
![image](https://user-images.githubusercontent.com/1218428/230777791-ee2440fb-af99-433b-a4da-cd347a535707.png)